src/linear_algebra/matrix/ldl.lean

```diff
@@ -0,0 +1,108 @@
+
+This file proves the LDL-decomposition of matricies: Any positive definite matrix `S` can be
+decomposed as `S = LDLᴴ` where `L` is a lower-triangular matrix and `D` is a diagonal matrix.
+
+## Main definitions
+
+ * `LDL.lower` is the lower triangular matrix `L`.
+ * `LDL.lower_inv` is the inverse of the lower triangular matrix `L`.
+ * `LDL.diag` is the diagonal matrix `D`.
+
+## Main result
+
+* `ldl_decomposition` states that any positive definite matrix can be decomposed as `LDLᴴ`.
+
+## TODO
+
+* Prove that `LDL.lower` is lower triangular from `LDL.lower_inv_triangular`.
+
+-/
+
+variables {𝕜 : Type*} [is_R_or_C 𝕜]
+  {n : Type*} [linear_order n] [is_well_order n (<)] [locally_finite_order_bot n]
+
+local notation `⟪`x`, `y`⟫` :=
+  @inner 𝕜 (n → 𝕜) (pi_Lp.inner_product_space (λ _, 𝕜)).to_has_inner x y
+
+open matrix
+open_locale matrix
+variables {S : matrix n n 𝕜} [fintype n] (hS : S.pos_def)
+
+/-- The inverse of the lower triangular matrix `L` of the LDL-decomposition. It is obtained by
+applying Gram-Schmidt-Orthogonalization w.r.t. the inner product induced by `Sᵀ` on the standard
+basis vectors `pi.basis_fun`. -/
+noncomputable def LDL.lower_inv : matrix n n 𝕜 :=
+  @gram_schmidt
+    𝕜 (n → 𝕜) _ (inner_product_space.of_matrix hS.transpose) n _ _ _ (λ k, pi.basis_fun 𝕜 n k)
+
+lemma LDL.lower_inv_gram_schmidt_basis :
+  LDL.lower_inv hS = ((pi.basis_fun 𝕜 n).to_matrix
+    (@gram_schmidt_basis 𝕜 (n → 𝕜) _
+    (inner_product_space.of_matrix hS.transpose) n _ _ _ (pi.basis_fun 𝕜 n)))ᵀ :=
+begin
+  ext i j,
+  rw [LDL.lower_inv, basis.coe_pi_basis_fun.to_matrix_eq_transpose, coe_gram_schmidt_basis],
+  refl
+end
+
+noncomputable instance LDL.invertible_lower_inv : invertible (LDL.lower_inv hS) :=
+begin
+  rw [LDL.lower_inv_gram_schmidt_basis],
+  haveI := basis.invertible_to_matrix (pi.basis_fun 𝕜 n)
+    (@gram_schmidt_basis 𝕜 (n → 𝕜) _ (inner_product_space.of_matrix hS.transpose)
+      n _ _ _ (pi.basis_fun 𝕜 n)),
+  apply_instance
+end
+
+lemma LDL.lower_inv_orthogonal {i j : n} (h₀ : i ≠ j) :
+  ⟪(LDL.lower_inv hS i), Sᵀ.mul_vec (LDL.lower_inv hS j)⟫ = 0 :=
+show @inner 𝕜 (n → 𝕜) (inner_product_space.of_matrix hS.transpose).to_has_inner
+    (LDL.lower_inv hS i)
+    (LDL.lower_inv hS j) = 0,
+by apply gram_schmidt_orthogonal _ _ h₀
+
+/-- The entries of the diagonal matrix `D` of the LDL decomposition. -/
+noncomputable def LDL.diag_entries : n → 𝕜 :=
+  λ i, ⟪star (LDL.lower_inv hS i), S.mul_vec (star (LDL.lower_inv hS i))⟫
+
+/-- The diagonal matrix `D` of the LDL decomposition. -/
+noncomputable def LDL.diag : matrix n n 𝕜 := matrix.diagonal (LDL.diag_entries hS)
+
+lemma LDL.lower_inv_triangular {i j : n} (hij : i < j) :
+  LDL.lower_inv hS i j = 0 :=
+by rw [← @gram_schmidt_triangular 𝕜 (n → 𝕜) _ (inner_product_space.of_matrix hS.transpose) n _ _ _
+    i j hij (pi.basis_fun 𝕜 n), pi.basis_fun_repr, LDL.lower_inv]
+
+lemma ldl_decomposition₀ : LDL.diag hS = LDL.lower_inv hS ⬝ S ⬝ (LDL.lower_inv hS)ᴴ :=
```

Comment:
Since there are multiple ways to phrase the theorem, let's name the lemma based on the result and give the common name in the docstring.
```suggestion
/-- Inverse statement of **LDL decomposition**: we can conjugate a positive definite matrix
by some lower triangular matrix and get a diagonal matrix. -/
lemma LDL.diag_eq_lower_inv_conj : LDL.diag hS = LDL.lower_inv hS ⬝ S ⬝ (LDL.lower_inv hS)ᴴ :=
```